File: built-in plugins dsp/mb-dist-tube.cpp
/**
 * MB Tube Distortion
 * Category : effect
 * Type     : distortion
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Warm tube saturation
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_DIST_TUBE_H
#define MB_DIST_TUBE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbDistTube : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-dist-tube";
    static constexpr const char* PLUGIN_NAME    = "MB Tube Distortion";
    static constexpr const char* PLUGIN_TYPE    = "distortion";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float drive = 0.3f;  // range [0, 1]
    float tone = 0.5f;  // range [0, 1]
    float mix = 1f;  // range [0, 1]
    };

    MbDistTube() = default;
    ~MbDistTube() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.drive = std::clamp(params.drive, 0f, 1f);
        params.tone = std::clamp(params.tone, 0f, 1f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Tube Distortion
        return input;
    }
};

#endif // MB_DIST_TUBE_H

File: built-in plugins dsp/mb-dist-tube.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbDistTubePlugin: PluginDefinition = { id: 'mb-dist-tube', slug: 'mb-dist-tube', name: 'MB Tube Distortion', category: 'effect', type: 'distortion', version: '1.0.0', description: 'Warm tube saturation', author: 'Max Booster', grade: 'A', parameters: [{ id: 'drive', name: 'Drive', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true }, { id: 'tone', name: 'Tone', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 1, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { drive: 0.3, tone: 0.5, mix: 1 } };

export default MbDistTubePlugin;
